scripts/Dockerfile

```diff
@@ -1,21 +1,32 @@
+FROM golang:1.11.4-alpine AS gobase
+RUN apk add --no-cache git bash
+
+# Compile metrictank
+ENV MT_SRC_DIR $GOPATH/src/github.com/grafana/metrictank
+RUN mkdir -p $MT_SRC_DIR
+ADD . $MT_SRC_DIR
+RUN $MT_SRC_DIR/scripts/build.sh
+RUN cp -r $MT_SRC_DIR/build /tmp/build
```

Comment:
why do we need `/tmp/build` ? what's wrong with `$MT_SRC_DIR/build` ?

Reply:
also,it would be nice to, when we're building through docker, to keep the freshly built binaries also in the local (host) build dir, in addition to inside of the container. so maybe we should create a volume map for the build dir, and copy the binaries in there.

or actually, we could volume mount the entire MT_SRC_DIR I think? and then copy the build dir from that into an inside-of-the-docker-image build dir?

Reply:
> why do we need `/tmp/build` ? what's wrong with `$MT_SRC_DIR/build` ?

The environment variables are local to each build stage, in other words $MT_SRC_DIR is not available anymore in the next build stage. Instead we use /tmp/build

Reply:
> also,it would be nice to, when we're building through docker, to keep the freshly built binaries also in the local (host) build dir, in addition to inside of the container. so maybe we should create a volume map for the build dir, and copy the binaries in there.
> 
> or actually, we could volume mount the entire MT_SRC_DIR I think? and then copy the build dir from that into an inside-of-the-docker-image build dir?

hmmm, so the use case would be we'd like to retrieve the mt binary to do some tests locally or something. I suppose this is not a very common scenario cause usually the tests would better be performed on the Docker image that is the final artifact of the build process.
However retrieving the mt binaries is already doable by using `docker cp`, which is slightly less practical but assuming it's not a very common scenario, I would guess it's enough, no?

Reply:
> The environment variables are local to each build stage, in other words $MT_SRC_DIR is not available anymore in the next build stage. Instead we use /tmp/build

the environment variable may cease to exist, but the directory would still be there, no?


> hmmm, so the use case would be we'd like to retrieve the mt binary to do some tests locally or something. I suppose this is not a very common scenario cause usually the tests would better be performed on the Docker image that is the final artifact of the build process.
However retrieving the mt binaries is already doable by using docker cp, which is slightly less practical but assuming it's not a very common scenario, I would guess it's enough, no?

I think there is a case to be made for having the binaries on the host (and don't consider the docker image to be the end of the build process) (otherwise why are we keeping `make bin` dockerless?)
my point is, if we're going to build fresh binaries from the current source in docker, than they will definitely be as up to date, or more up to date than the binaries we already have on the host. so we may as well copy them over, it's a convenient way to make sure we have the latest binaries locally as well. definitely more convenient to telling users to run `docker cp`.
the counterpoint to my argument is that now you use diskspace twice for the binaries (once the image, once in the local dir)

Reply:
> > The environment variables are local to each build stage, in other words $MT_SRC_DIR is not available anymore in the next build stage. Instead we use /tmp/build
> 
> the environment variable may cease to exist, but the directory would still be there, no?
> 

yes but we don't know the path of that directory

Reply:
> > hmmm, so the use case would be we'd like to retrieve the mt binary to do some tests locally or something. I suppose this is not a very common scenario cause usually the tests would better be performed on the Docker image that is the final artifact of the build process.
> > However retrieving the mt binaries is already doable by using docker cp, which is slightly less practical but assuming it's not a very common scenario, I would guess it's enough, no?
> 
> I think there is a case to be made for having the binaries on the host (and don't consider the docker image to be the end of the build process) (otherwise why are we keeping `make bin` dockerless?)
> my point is, if we're going to build fresh binaries from the current source in docker, than they will definitely be as up to date, or more up to date than the binaries we already have on the host. so we may as well copy them over, it's a convenient way to make sure we have the latest binaries locally as well. definitely more convenient to telling users to run `docker cp`.
> the counterpoint to my argument is that now you use diskspace twice for the binaries (once the image, once in the local dir)

We can copy them over automatically but I don't think that in practice devs would use those very often. As 'proof' I would say that I don't recall anyone asking from them until now. In other words, in what case would the devs use those exactly?

Reply:
right. it is quite rare to run the binaries directly without docker.
but note that many of the docker stacks volume-mount the binary from the host build directory! 
this means it's very quick to rebuild (only the main binary, not all the tools), and restart the container, rather than having to wait for all tools to build. it leads to a quick code-iteration cycle.
also if the binary doesn't exist in the local build dir, the container won't run and create root owned files on the host directory. quite annoying.
so either we:
1) remove the `- ../../build/metrictank:/usr/bin/metrictank` volume from all docker compose files, and always require a complete image rebuild to be able to retest after making code changes
2) tell users to always run the build script locally to create local binaries to use some of these docker stacks
3) automatically copy the binaries from the container to the host. maybe not so commonly needed, but i think it is a handy behavior for those cases where it's needed.

Reply:
> yes but we don't know the path of that directory

sure we do. it's `$GOPATH/src/github.com/grafana/metrictank`. BTW where is `$GOPATH` set? is it always set in the golang container?

Reply:
> > yes but we don't know the path of that directory
> 
> sure we do. it's `$GOPATH/src/github.com/grafana/metrictank`. BTW where is `$GOPATH` set? is it always set in the golang container?

`$GOPATH` depends on the docker image used in the stage. For example it's `/go` in `golang:1.11.4-alpine` but it's not set in `alpine:3.10.0` which is why we cannot use `$GOPATH/src/github.com/grafana/metrictank`

Reply:
Ok so then we know the files live in `/go/src/github.com/grafana/metrictank`

Reply:
> Ok so then we know the files live in `/go/src/github.com/grafana/metrictank`

Not if $GOPATH in the first image is not `/go`